docs/contrib/build.rst

```diff
@@ -0,0 +1,159 @@
+
+This section gives some basic information and tips about the build system. The
+``sbt`` build system is quite complex and effectively brings together all the
+components of Scala Native. The ``build.sbt`` file is at the root of the project
+along with the sub-projects that make up the system.
+
+Overview
+--------------------------------
+In order to effectively work with Scala Native, a knowledge of the build system
+is very helpful. In general the code is built and published to your local Ivy
+repository using the `sbt` `publishLocal` command so that other components in the
+system can depend on each other via normal `sbt` dependencies. Although the
+``build.sbt`` file and other code in the system is the way to learn the system
+thoroughly, the following sections will give information that should be helpful
+to get started.
+
+The build has roughly four groups of sub-projects as follows:
+
+1.  The Native Scala Compiler plugin and libraries. Each of these projects depend
+    on the next project in the list.
+
+    - `nscplugin`
+
+    - `nativelib`
+
+    - `javalib`
+
+    - `auxlib`
+
+    - `scalalib`
+
+2.  The Scala Native plugin and dependencies (directory names are in parentheses).
+
+    - `sbtScalaNative (sbt-scala-native)`
+
+    - `tools`
+
+    - `nir`
+
+    - `util`
+
+3.  The Scala Native test interface and dependencies.
+
+    - `testInterface (test-interface)`
+
+    - `testInterfaceSerialization (test-interface-serialization)`
+
+    - `testInterfaceSbtDefs (test-interface-sbt-defs)`
+
+4.  Tests and benchmarks (no dependencies on each other).
+
+    - `tests (unit-tests)`
+
+    - `tools` This has tests within the project
+
+    - `(scripted-tests)`
+
+    - `benchmarks`
+
+Each of the groups above also depend on the previous group being compiled and
+published locally. The sbt plugin ``sbtScalaNative`` is used inside Scala Native
+exactly as it is used in a project using Scala Native. The plugin is needed
+by the `testInterface` and also the `tests` that use the `testInterface`
+to compile native code.
+
+Building Scala Native
+---------------------
+Once you have cloned Scala Native from git, `cd` into the base directory. Inside
+this directory is the `build.sbt` file which is used to build Scala Native. This
+file has `sbt` command aliases which are used to help build the system. In order
+to build Scala Native for the first time you should run the following commands:
+
+.. code-block:: text
+
+    $ sbt
+    > rebuild
+
+If you want to run all the tests and benchmarks, which take awhile you can run
+the `test-all` command after the systems builds.
+
+Normal development workflow
+---------------------------
+Let us suppose that you wish to work on the `javalib` project to add some code
+or fix a bug. Once you make a change to the code, run the following command
+at the sbt prompt to compile the code and run the tests:
+
+.. code-block:: text
+
+    > javalib/publishLocal
+    > tests/test
```

Comment:
if there's a change in codegen, is `rebuild` the only option?

Reply:
This was suppose to cover the basics. The next section talks about the codegen but I have never worked on that myself. http://www.scala-native.org/en/latest/contrib/compiler.html

Reply:
If you change anything in `tools` (linker, optimizer, codegen), you need to `reload`, not `rebuild`. It's possible because we textually include code of sbt plugin and the toolchain. 

If you change `nscplugin`, `rebuild` is the only option.

Reply:
I included this in the docs as well.